src/dbnode/storage/bootstrap.go

```diff
@@ -70,42 +70,154 @@ const (
 
 type bootstrapFn func() error
 
-type bootstrapManager struct {
-	sync.RWMutex
-
-	database                    database
-	mediator                    databaseMediator
+type instrumentation struct {
 	opts                        Options
 	log                         *zap.Logger
-	bootstrapFn                 bootstrapFn
 	nowFn                       clock.NowFn
 	sleepFn                     sleepFn
-	processProvider             bootstrap.ProcessProvider
-	state                       BootstrapState
-	hasPending                  bool
 	status                      tally.Gauge
 	bootstrapDuration           tally.Timer
+	bootstrapNamespacesDuration tally.Timer
 	durableStatus               tally.Gauge
 	lastBootstrapCompletionTime xtime.UnixNano
+	start                       time.Time
+	startNamespaces             time.Time
+	logFields                   []zapcore.Field
+}
+
+func (i *instrumentation) bootstrapFnFailed(retry int) {
+	i.log.Warn("retrying bootstrap after backoff",
+		zap.Duration("backoff", bootstrapRetryInterval),
+		zap.Int("numRetries", retry+1))
+	i.sleepFn(bootstrapRetryInterval)
+}
+
+func (i *instrumentation) bootstrapPreparing() {
+	i.start = i.nowFn()
+	i.log.Info("bootstrap prepare")
+}
+
+func (i *instrumentation) bootstrapPrepareFailed(err error) {
+	i.log.Error("bootstrap prepare failed", zap.Error(err))
+}
+
+func (i *instrumentation) bootstrapStarted(shards int) {
+	i.logFields = []zapcore.Field{
+		zap.Int("numShards", shards),
+	}
+	i.log.Info("bootstrap started", i.logFields...)
+}
+
+func (i *instrumentation) bootstrapSucceeded() {
+	bootstrapDuration := i.nowFn().Sub(i.start)
+	i.bootstrapDuration.Record(bootstrapDuration)
+	i.logFields = append(i.logFields, zap.Duration("bootstrapDuration", bootstrapDuration))
+	i.log.Info("bootstrap succeeded, marking namespaces complete", i.logFields...)
+}
+
+func (i *instrumentation) bootstrapFailed(err error) {
+	bootstrapDuration := i.nowFn().Sub(i.start)
+	i.bootstrapDuration.Record(bootstrapDuration)
+	i.logFields = append(i.logFields, zap.Duration("bootstrapDuration", bootstrapDuration))
+	i.log.Error("bootstrap failed", append(i.logFields, zap.Error(err))...)
+}
+
+func (i *instrumentation) bootstrapNamespaceFailed(err error, namespaceID string) {
```

Comment:
Nit (for a stronger type and cleaner call site):
```suggestion
func (i *instrumentation) bootstrapNamespaceFailed(err error, namespaceID ident.ID) {
```